### PR TITLE
Faster http protocol, part 1 (and maybe faster Foolscap too, while we're at it)

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3939

Improvements:

1. For `allmydata.test.test_system.HTTPSystemTest.test_filesystem`, number of writes during immutable upload goes from 530 to 60. So that should be nice both for Foolscap and HTTP, though more so for the latter since requests have higher bandwidth overhead.
2. Update batching size limit that was last updated 15 years ago or something. This should enable much faster uploads for people with decent uplink bandwidth. Previously it sent ~50KB at a time, and I _think_ (but haven't strongly validated) it was waiting for response before continuing. Now it sends 1MB at at a time, reducing the impact of latency. Ideally this would be dynamic based on actual network conditions: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3787
3. Simplified some code by switching batching (so-called "pipelining") from RPC calls to the data level.

This did not reduce _all_ the protocol chattiness. As a proxy measure, non-persistent HTTPS is still a big slowdown compared to persistent HTTPS, though it slightly improved over the main branch. I will try to continue improvements in follow-up PRs.

### CODE REVIEW

**Review by exarkun — APPROVED** [review]
Thanks.  This looks like a nice simplification and hopefully speedup.  There are a few minor comments inline.  Please address to your satisfaction and merge.

Tangentially, is it worth talking about doing some performance measurement work soon?

File: src/allmydata/immutable/encode.py [hunk 1]
```diff
@@ -262,6 +262,8 @@ def start(self):
 
         d.addCallback(lambda res: self.finish_hashing())
 
+        # These calls have to happen in order, and waiting for previous one
+        # also ensures backpressure:
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/immutable/encode.py:266]: The layout code now assumes that writes are appending, essentially. Until recently they could happen in any order, and holes were OK (and there was a hole, in fact).

File: src/allmydata/immutable/layout.py [hunk 2]
```diff
@@ -107,19 +104,54 @@ def make_write_bucket_proxy(rref, server,
...
+    def queue_write(self, data: bytes) -> bool:
+        """
+        Queue a write.  If the result is ``False``, no further action is needed
+        for now.  If the result is some ``True``, it's time to call ``flush()``
+        and do a real write.
+        """
+        self._to_write.write(data)
+        return len(self._to_write.getbuffer()) >= self._batch_size
+
+    def flush(self) -> tuple[int, bytes]:
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/immutable/layout.py:126]: `from __future__ import annotations` enables this on older Python.

[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/immutable/layout.py:126]: E.g.

```
(venv37) $ mypy example.py
example.py:4: error: Incompatible types in assignment (expression has type "Tuple[int, str]", variable has type "Tuple[int, int]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
(venv37) $ cat example.py 
from __future__ import annotations

x: tuple[int, int]
x = (1, "a")
y = (2, 3)
```

File: newsfragments/3939.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+Uploading immutables will now use more bandwidth, which should allow for faster uploads in many cases.
```
[the remark this answers is not in the document]
Comment (itamarst) [reply newsfragments/3939.bugfix:1]: Good point

File: src/allmydata/immutable/layout.py [hunk 4]
```diff
@@ -237,29 +271,41 @@ def put_share_hashes(self, sharehashes):
...
+        return self._rref.callRemote("write", offset, data)
 
     def close(self):
-        assert self._written_bytes == self.get_allocated_size(), f"{self._written_bytes} != {self.get_allocated_size()}"
-        d = self._pipeline.add(0, self._rref.callRemote, "close")
-        d.addCallback(lambda ign: self._pipeline.flush())
+        assert self._write_buffer.get_total_bytes_queued() == self.get_allocated_size(), (
+            f"{self._written_buffer.get_total_bytes_queued()} != {self.get_allocated_size()}"
+        )
+        d = self._actually_write()
```
[the remark this answers is not in the document]
Comment (itamarst) [reply src/allmydata/immutable/layout.py:307]: I made it not send those.

**Pull request merged**